extern/sector-storage/piece_provider.go

```diff
@@ -0,0 +1,134 @@
+//
+// Will return a nil reader if the piece does NOT exist in any unsealed file/there is not unsealed file for the given sector on any of the workers.
+func (p *pieceProvider) tryReadUnsealedPiece(ctx context.Context, sector storage.SectorRef, offset storiface.UnpaddedByteIndex, size abi.UnpaddedPieceSize) (io.ReadCloser, context.CancelFunc, error) {
+	// acquire a lock purely for reading unsealed sectors
+	ctx, cancel := context.WithCancel(ctx)
+	if err := p.index.StorageLock(ctx, sector.ID, storiface.FTUnsealed, storiface.FTNone); err != nil {
+		cancel()
+		return nil, nil, xerrors.Errorf("acquiring read sector lock: %w", err)
+	}
+
+	r, err := p.storage.Reader(ctx, sector, abi.PaddedPieceSize(offset.Padded()), size.Padded())
+	if err != nil {
+		cancel()
+		return nil, nil, err
+	}
+	if r == nil {
+		cancel()
+	}
+
+	return r, cancel, nil
+}
+
+// ReadPiece is used to read an Unsealed piece at the given offset and of the given size from a Sector
+// If an Unsealed sector file exists with the Piece Unsealed in it, we'll use that for the read.
+// Otherwise, we will Unseal a Sealed sector file for the given sector and read the Unsealed piece from it.
+func (p *pieceProvider) ReadPiece(ctx context.Context, sector storage.SectorRef, offset storiface.UnpaddedByteIndex, size abi.UnpaddedPieceSize, ticket abi.SealRandomness, unsealed cid.Cid) (io.ReadCloser, bool, error) {
+	if err := offset.Valid(); err != nil {
+		return nil, false, xerrors.Errorf("offset is not valid: %w", err)
+	}
+	if err := size.Validate(); err != nil {
+		return nil, false, xerrors.Errorf("size is not a valid piece size: %w", err)
+	}
+
+	r, unlock, err := p.tryReadUnsealedPiece(ctx, sector, offset, size)
+	if xerrors.Is(err, storiface.ErrSectorNotFound) {
+		log.Debugf("no unsealed sector file with unsealed piece, sector=%+v, offset=%d, size=%d", sector, offset, size)
+		err = nil
+	}
+	if err != nil {
+		return nil, false, err
+	}
+
+	var uns bool
+	if r == nil {
+		uns = true
+		commd := &unsealed
+		if unsealed == cid.Undef {
+			commd = nil
+		}
+		if err := p.uns.SectorsUnsealPiece(ctx, sector, offset, size, ticket, commd); err != nil {
+			return nil, false, xerrors.Errorf("unsealing piece: %w", err)
+		}
+
+		log.Debugf("unsealed a sector file to read the piece, sector=%+v, offset=%d, size=%d", sector, offset, size)
+
+		r, unlock, err = p.tryReadUnsealedPiece(ctx, sector, offset, size)
+		if err != nil {
+			return nil, true, xerrors.Errorf("read after unsealing: %w", err)
+		}
+		if r == nil {
+			return nil, true, xerrors.Errorf("got no reader after unsealing piece")
+		}
+		log.Debugf("got a reader to read unsealed piece, sector=%+v, offset=%d, size=%d", sector, offset, size)
+	} else {
+		log.Debugf("unsealed piece already exists, no need to unseal, sector=%+v, offset=%d, size=%d", sector, offset, size)
+	}
+
+	upr, err := fr32.NewUnpadReader(r, size.Padded())
+	if err != nil {
+		return nil, uns, xerrors.Errorf("creating unpadded reader: %w", err)
```

Comment:
Do we need to call unlock() before returning the error here?

Reply:
Great catch !